solarforecastarbiter/power.py

```diff
@@ -0,0 +1,167 @@
+"""
+Calculate AC power and modeling intermediates from system metadata,
+times, and weather data.
+
+Steps are:
+
+1. Calculate solar position using solar_position
+2. If not already known, calculate 3 irradiance components from measured
+   GHI using irradiance_components or modeled clear sky using clearsky.
+3. calculate_poa_effective
+4. calculate_power
+"""
+
+import pvlib
+
+from solarforecastarbiter import datamodel
+
+
+def solar_position(site, times):
+    """Wraps :func:`pvlib.solarposition.spa_python`"""
+    solpos = pvlib.solarposition.spa_python(times, site.latitude,
+                                            site.longitude)
+    return solpos
+
+
+def irradiance_components(ghi, solar_position):
```

Comment:
I don't understand why a dataclass with system metadata is ok but a dataframe for each of the closely related parameters is not ok.

Zenith because that's what erbs wants.

Reply:
yeah, i also thought we were only going to use the attributes instead of passing the dataclasses, but I guess they are convenience functions. At least with the metadata, the attributes are defined in solararbiter core and can't change.

Reply:
Rewrote functions to use series input. 

The solar position and clear sky wrappers still return the pvlib DataFrames. We could unpack them at the cost of risking mixing up the return order.

I can rewrite to pass attributes next. Can we add methods to the dataclasses for convenience? They would wrap `ghi_dni_dhi_to_power` and `ghi_to_power`

Reply:
adding methods would go against the whole functional approach we're taking. I would allow passing the metadata object over adding methods to it.

Reply:
`func(*metadata, *data)` or `func(*data, *metadata)`?

e.g.

```python
def calculate_solar_position(latitude, longitude, elevation, times):
def calculate_clearsky(latitude, longitude, elevation, apparent_zenith):
# or 
def calculate_solar_position(times, latitude, longitude, elevation):
def calculate_clearsky(apparent_zenith, latitude, longitude, elevation):
```

Reply:
metadata first in latest commit. only the aoi and final convenience functions take the system. I don't know how to do it better given the different metadata for trackers and fixed systems.

Reply:
In the current version, `system.modeling_parameters` could be replaced with `modeling_parameters` everywhere in the module.